File: modules/expo-battery/android/src/main/java/expo/modules/battery/ExpoBatteryModuleSharedObject.kt
package expo.modules.battery

import expo.modules.kotlin.AppContext
import expo.modules.kotlin.sharedobjects.SharedObject

class ExpoBatteryModuleSharedObject(appContext: AppContext) : SharedObject(appContext) {
  var count: Int = 0

  override fun sharedObjectDidRelease() {
    super.sharedObjectDidRelease()
  }
}

File: modules/expo-battery/src/ExpoBatteryModuleSharedObject.ts
import { SharedObject, useReleasingSharedObject } from 'expo-modules-core';

import ExpoBatteryModule from './ExpoBatteryModule';

export declare class ExpoBatteryModuleSharedObject extends SharedObject {
  count: number;
}

/**
 * Creates a new ExpoBatteryModuleSharedObject instance.
 * You are responsible for releasing it from memory by calling `release()` when done.
 */
export function createExpoBatteryModuleSharedObject(): ExpoBatteryModuleSharedObject {
  return new ExpoBatteryModule.ExpoBatteryModuleSharedObject();
}

/**
 * A hook that creates a ExpoBatteryModuleSharedObject instance and automatically
 * releases it when the component unmounts.
 */
export function useExpoBatteryModuleSharedObject(): ExpoBatteryModuleSharedObject {
  return useReleasingSharedObject(() => new ExpoBatteryModule.ExpoBatteryModuleSharedObject(), []);
}
